File: sdk/go/pelisearch/integration_test.go
package pelisearch_test

import (
	"context"
	"os"
	"testing"
	"time"

	"github.com/Open-Collective-Labs/Peli-search/sdk/go/pelisearch"
)

func testURL() string {
	if u := os.Getenv("PELISEARCH_TEST_URL"); u != "" {
		return u
	}
	return "http://127.0.0.1:7700"
}

func TestIntegration(t *testing.T) {
	client := pelisearch.NewClientFromURL(testURL())
	ctx := context.Background()
	const index = "sdk_go_test"

	_ = client.DeleteIndex(ctx, index)

	if err := client.Health(ctx); err != nil {
		t.Fatalf("health: %v", err)
	}

	if _, err := client.CreateIndex(ctx, index); err != nil {
		t.Fatalf("create index: %v", err)
	}

	indexes, err := client.ListIndexes(ctx)
	if err != nil {
		t.Fatalf("list indexes: %v", err)
	}
	found := false
	for _, name := range indexes {
		if name == index {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("index %q not in list", index)
	}

	if _, err := client.AddDocument(ctx, index, "d1", map[string]interface{}{
		"title":    "Wireless Mouse",
		"category": "electronics",
		"price":    29.99,
	}); err != nil {
		t.Fatalf("add document: %v", err)
	}

	if _, err := client.BulkAddDocuments(ctx, index, []map[string]interface{}{
		{"id": "d2", "fields": map[string]interface{}{"title": "Keyboard", "category": "electronics"}},
	}); err != nil {
		t.Fatalf("bulk add: %v", err)
	}

	q := "mouse"
	results, err := client.Search(ctx, index, &pelisearch.SearchRequest{Q: &q})
	if err != nil {
		t.Fatalf("search: %v", err)
	}
	if len(results.Hits) == 0 {
		t.Fatal("expected search hits")
	}
	if results.Total < len(results.Hits) {
		t.Fatal("total should be at least hits length")
	}
	for _, hit := range results.Hits {
		if hit.Index == "" {
			t.Fatal("hit missing index")
		}
	}

	results, err = client.Search(ctx, index, &pelisearch.SearchRequest{
		Query: map[string]interface{}{"match": map[string]string{"title": "keyboard"}},
	})
	if err != nil {
		t.Fatalf("dsl search: %v", err)
	}
	if len(results.Hits) == 0 {
		t.Fatal("expected dsl search hits")
	}

	results, err = client.Search(ctx, index, &pelisearch.SearchRequest{
		Q:    strPtr("mouse"),
		From: intPtr(0),
		Size: intPtr(1),
	})
	if err != nil {
		t.Fatalf("paged search: %v", err)
	}
	if len(results.Hits) > 1 {
		t.Fatal("expected at most 1 hit with size=1")
	}

	if err := client.DeleteIndex(ctx, index); err != nil {
		t.Fatalf("delete index: %v", err)
	}
}

func strPtr(s string) *string { return &s }

func intPtr(i int) *int { return &i }

func TestContextTimeout(t *testing.T) {
	client := pelisearch.NewClientFromURL(testURL())
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := client.Health(ctx); err != nil {
		t.Fatalf("health with context: %v", err)
	}
}

func TestMissingIndexError(t *testing.T) {
	client := pelisearch.NewClientFromURL(testURL())
	_, err := client.GetIndex(context.Background(), "nonexistent_sdk_index")
	if err == nil {
		t.Fatal("expected error")
	}
	if _, ok := err.(*pelisearch.APIError); !ok {
		t.Fatalf("expected APIError, got %T", err)
	}
}

File: sdk/tests/javascript/integration.test.ts
import { beforeAll, describe, expect, it } from "vitest"
import { PeliSearchClient, PeliSearchError } from "@pelisearch/client"

const baseUrl = process.env.PELISEARCH_TEST_URL ?? "http://127.0.0.1:7700"
const client = new PeliSearchClient({ host: baseUrl })
const INDEX = "sdk_js_test"

async function resetIndex(name: string) {
  try {
    await client.deleteIndex(name)
  } catch {
    // index may not exist
  }
  await client.createIndex(name)
}

beforeAll(async () => {
  await client.health()
  await resetIndex(INDEX)
})

describe("index management", () => {
  it("creates, lists, gets, and deletes indexes", async () => {
    const temp = "sdk_js_index_crud"
    await resetIndex(temp)

    const names = await client.listIndexes()
    expect(names).toContain(temp)

    const info = await client.getIndex(temp)
    expect(info.name).toBe(temp)
    expect(info.document_count).toBe(0)

    await client.deleteIndex(temp)
    const after = await client.listIndexes()
    expect(after).not.toContain(temp)
  })

  it("returns typed errors for missing indexes", async () => {
    try {
      await client.getIndex("nonexistent_sdk_index")
      expect.fail("expected error")
    } catch (err) {
      expect(err).toBeInstanceOf(PeliSearchError)
      expect((err as PeliSearchError).status).toBe(404)
    }
  })
})

describe("documents", () => {
  beforeAll(async () => {
    await resetIndex(INDEX)
  })

  it("adds, gets, bulk adds, and deletes documents", async () => {
    await client.addDocument(INDEX, "d1", { title: "Mouse", category: "electronics", price: 29.99 })
    const doc = await client.getDocument(INDEX, "d1")
    expect(doc.fields?.title ?? doc.title).toBeTruthy()

    const bulk = await client.bulkAddDocuments(INDEX, [
      { id: "d2", fields: { title: "Keyboard", category: "electronics", price: 89.99 } },
    ])
    expect(bulk.documents[0].status).toBe("created")

    await client.deleteDocument(INDEX, "d1")
  })
})

describe("search", () => {
  beforeAll(async () => {
    await resetIndex(INDEX)
    await client.bulkAddDocuments(INDEX, [
      { id: "p1", fields: { title: "Wireless Mouse", category: "electronics", price: 29.99 } },
      { id: "p2", fields: { title: "Mechanical Keyboard", category: "electronics", price: 89.99 } },
    ])
  })

  it("searches with legacy q", async () => {
    const results = await client.search(INDEX, { q: "mouse" })
    expect(results.hits.length).toBeGreaterThan(0)
    expect(results.total).toBeGreaterThan(0)
    for (const hit of results.hits) {
      expect(hit.index).toBe(INDEX)
      expect(hit.document_id).toBeTruthy()
      expect(hit.score).toBeGreaterThanOrEqual(0)
    }
  })

  it("searches with DSL match", async () => {
    const results = await client.search(INDEX, {
      query: { match: { title: "keyboard" } },
    })
    expect(results.hits.length).toBeGreaterThan(0)
  })

  it("supports pagination", async () => {
    const results = await client.search(INDEX, { q: "mouse keyboard", from: 0, size: 1 })
    expect(results.hits.length).toBeLessThanOrEqual(1)
  })
})
